runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -12418,6 +12418,905 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   auto scalarLoadOp = isLatin1 ? TR::InstOpCode::lbzx : TR::InstOpCode::lhzx;
+
+   TR::Register *arrayObjReg     = cg->evaluate(node->getChild(0));
+   TR::Register *arrayLenReg     = cg->evaluate(node->getChild(1));
+   TR::Register *subStringObjReg = cg->evaluate(node->getChild(2));
+   TR::Register *strCountReg     = cg->evaluate(node->getChild(3));
+   TR::Register *fromIndexReg    = cg->evaluate(node->getChild(4));
+
+   TR::Register *subStringLenReg = cg->allocateRegister();
+   TR::Register *arrayAddressReg = cg->allocateRegister();
+   TR::Register *resultReg       = cg->allocateRegister();
+
+   TR::Register *vec1Reg            = cg->allocateRegister(TR_VRF);
+   TR::Register *vec2Reg            = cg->allocateRegister(TR_VRF);
+   TR::Register *targetVectorReg    = cg->allocateRegister(TR_VRF);
+   TR::Register *targetVectorNotReg = cg->allocateRegister(TR_VRF);
+   TR::Register *searchVectorReg    = cg->allocateRegister(TR_VRF);
+   TR::Register *permuteVectorReg   = cg->allocateRegister(TR_VRF);
+   TR::Register *maskVectorReg      = cg->allocateRegister(TR_VRF);
+
+   TR::Register *cr0Reg = cg->allocateRegister(TR_CCR);
+   TR::Register *cr6Reg = cg->allocateRegister(TR_CCR);
+
+   TR_PPCScratchRegisterManager *srm = cg->generateScratchRegisterManager();
+
+   TR::LabelSymbol *startLabel            = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundEarlyLabel       = generateLabelSymbol(cg);
+   TR::LabelSymbol *notFoundEarlyLabel    = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundEndVectorLabel   = generateLabelSymbol(cg);
+   TR::LabelSymbol *notSmallLabel         = generateLabelSymbol(cg);
+   TR::LabelSymbol *firstContinueLabel    = generateLabelSymbol(cg);
+   TR::LabelSymbol *vectorLoopLabel       = generateLabelSymbol(cg);
+   TR::LabelSymbol *vectorContinueLabel   = generateLabelSymbol(cg);
+   TR::LabelSymbol *residueLabel          = generateLabelSymbol(cg);
+   TR::LabelSymbol *notFoundLabel         = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundFirstVectorLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundVectorLabel      = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundLabel            = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundExactLabel       = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel              = generateLabelSymbol(cg);
+
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+      {
+      /* If offset + number of bytes in the substring is greater than the number of bytes in the main string it can't be found. In this case jump to notFoundLabel to return -1. */
+      TR::Register *minStrByteCountReg = srm->findOrCreateScratchRegister();
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, minStrByteCountReg, fromIndexReg, strCountReg);
+      if (!isLatin1)
```

Comment:
isn't it weird that arrayLen is represented in terms of byte, while fromIndex and strCount are expected to be represented in terms of element?

if all are consistently meaning element, you don't need to do this doubling, that is.

Reply:
The line right after this was an error. arrayLen, strCount and fromIndex all contain a element count or index to begin with. So the value does not need to be adjusted even under UTF16. I've pushed a change to fix this.